src/dlsnode/storage/util/InputBuffer.d

```diff
@@ -145,6 +201,68 @@ struct InputBuffer
         return this.data_in_chunk - this.position_in_chunk;
     }
 
+    /***************************************************************************
+
+        Tries to read data in non-blocking manner. If the requested data is
+        available in the internal buffer, this method would return immediately,
+        filling the buffer with the requested data. Otherwise, provided method
+        async_read_data will be called, expecting to fill the internal buffer 
+        and call asyncReadCompleted to let InputBuffer update it's state with the data
+        actually read from the file. Either way, this method will return
+        *immediately* and user should check if the read has completed inspecting
+        the Future object.
+
+        Params:
+            promise = promise to fulfil
+            async_read_data = delegate to read the data and let the input buffer know
+                        when it's done (via asyncReadCompleted parameter).
+
+    ***************************************************************************/
+
+    public void asyncReadData (ref Promise promise,
+            void delegate(void[] dest,
+                          void delegate(ssize_t) asyncReadCompleted) async_read_data)
+    {
+        // If we don't need to read anything
+        if (promise.dataMissing() == 0)
+        {
+            promise.fulfilled(false);
+            return;
+        }
+
+        // Do we have enough data already in the buffer?
+        if (this.remainingInBuffer() >= promise.dataMissing)
+        {
+            this.consumeFromInternalBuffer(promise.dataMissing(), &promise.fillResult);
+            promise.fulfilled(false);
+            return;
+        }
+        else if (this.feof_occured)
+        {
+            // Copy what we can and return that.
+            auto remaining = this.remainingInBuffer();
+            this.consumeFromInternalBuffer(remaining, &promise.fillResult);
+            promise.fulfilled(true);
+            return;
+        }
+
+        // There's not enough data in the buffer? Let's move the remaining of the
+        // data to the front of the buffer, and refill the buffer with the new data
+        auto remaining = this.moveRemainingToFront();
+
+        // There's not enough data in the buffer? At this point we need
+        // to refill the buffer
+        if (this.buffer.length < promise.dataMissing)
```

Comment:
Isn't this condition always true at this point?

Reply:
Ah I think the comment was what confused me. This step isn't about refilling the buffer, it's about resizing it.

Reply:
Ah, yes, the comment is wrong. Will fix.